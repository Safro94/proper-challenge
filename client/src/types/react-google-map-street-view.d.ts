declare module 'react-google-map-street-view';
