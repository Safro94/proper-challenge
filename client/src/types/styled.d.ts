import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				black: string;
				gray: string;
				white: string;
				danger: string;
			};
			primary: IPalette;
		};
	}
}
