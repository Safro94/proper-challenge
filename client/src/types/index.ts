export interface IProperty {
	id: number;
	address: string;
	size: string;
	rooms: number;
	utilities: string;
	tenantName: string;
}

export enum ButtonTypes {
	Primary = 'primary',
	Danger = 'danger',
}

export enum RequestStatus {
	Idle = 'idle',
	Pending = 'pending',
	Resolved = 'resolved',
	Rejected = 'rejected',
}
