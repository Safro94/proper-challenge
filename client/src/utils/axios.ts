import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);
instance.defaults.headers.common['x-api-key'] = process.env.REACT_APP_API_KEY;

export default instance;
