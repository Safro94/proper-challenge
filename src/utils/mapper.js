const mapProperties = properties =>
	properties.map(({ address, size, rooms, utilities, tenantName }) => ({
		address,
		size,
		rooms,
		utilities,
		tenantName,
	}));

module.exports = { mapProperties };
