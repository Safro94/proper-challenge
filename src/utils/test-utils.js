const mockReq = params => {
	const req = { ...params };
	return req;
};

const mockRes = params => {
	const res = { ...params };
	res.json = jest.fn().mockReturnValue(res);
	res.status = jest.fn().mockReturnValue(res);

	return res;
};

module.exports = {
	mockReq,
	mockRes,
};
