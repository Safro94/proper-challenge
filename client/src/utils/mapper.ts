import { IAddress, IDawaResponse } from '../types';

const mapDawaResponse = (data: IDawaResponse[]): IAddress[] => {
	return data.map(item => ({
		text: item.tekst,
		type: item.type,
	}));
};

export { mapDawaResponse };
