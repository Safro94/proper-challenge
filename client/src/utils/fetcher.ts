import { RequestMethods, RequestStatus } from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
	headers?: null | any;
}

interface IResponse {
	data: any;
	status: RequestStatus;
}

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
	headers = null,
}: IFetcher): Promise<IResponse> => {
	let data;
	let status = RequestStatus.Pending;

	try {
		const result = await axios[method](
			url,
			JSON.parse(headers),
			JSON.parse(body)
		);
		data = result.data;
		status = RequestStatus.Resolved;
	} catch (error) {
		status = RequestStatus.Rejected;
		throw error;
	}

	return { data, status };
};

export default fetcher;
