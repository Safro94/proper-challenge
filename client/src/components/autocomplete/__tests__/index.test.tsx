import { render, screen } from '../../../utils/test-utils';

import Autocomplete from '..';
import { IAddress } from '../../../types';
import userEvent from '@testing-library/user-event';

describe('Autocomplete', () => {
	const items = [{ text: 'Matias' }, { text: 'Martin' }, { text: 'Charles' }];
	const labelText = 'text';
	const filterProp = 'text';
	const onInputChange = () => {};
	const onChange = () => {};
	const placeholder = 'input';

	const Component = () => (
		<Autocomplete
			items={items}
			labelText={labelText}
			filterProp={filterProp}
			onInputChange={onInputChange}
			render={(getMenuProps: any, inputItems: IAddress[]) => (
				<div {...getMenuProps()}>
					<ul>
						{inputItems.map((item: IAddress) => (
							<li key={item.text} data-testid='option'>
								{item.text}
							</li>
						))}
					</ul>
				</div>
			)}
			placeholder={placeholder}
			onChange={onChange}
		/>
	);

	it('should render correctly', () => {
		const { container } = render(<Component />);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe('Autocomplete functionality', () => {
		beforeEach(() => {
			render(<Component />);
		});

		it('should filter the options while the user types', () => {
			const input = screen.getByPlaceholderText(/input/i);

			expect(screen.getByText(/text/i)).toBeInTheDocument();
			expect(input).toBeInTheDocument();

			expect(screen.getAllByTestId('option')).toHaveLength(3);

			userEvent.type(input, 'Ma');
			expect(screen.getAllByTestId('option')).toHaveLength(2);

			userEvent.type(input, 't');
			expect(screen.getAllByTestId('option')).toHaveLength(1);
		});
	});
});
