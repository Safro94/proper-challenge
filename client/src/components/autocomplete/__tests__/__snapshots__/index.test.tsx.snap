// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render correctly 1`] = `
<div
  class="sc-bdnxRM iWitlZ"
>
  <div
    class="sc-dlnjwi cSrDLl"
  >
    <label
      class="sc-eCApnc fahIwV"
      for="downshift-0-input"
      id="downshift-0-label"
    >
      text
    </label>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-0-menu"
      class="sc-gtsrHT gIlwKp"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-0-menu"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="sc-hKFxyN daVEHi"
        id="downshift-0-input"
        placeholder="input"
        value=""
      />
    </div>
  </div>
  <div
    aria-labelledby="downshift-0-label"
    id="downshift-0-menu"
    role="listbox"
  >
    <ul>
      <li
        data-testid="option"
      >
        Matias
      </li>
      <li
        data-testid="option"
      >
        Martin
      </li>
      <li
        data-testid="option"
      >
        Charles
      </li>
    </ul>
  </div>
</div>
`;
